aten/src/ATen/native/cuda/TensorFactories.cu

```diff
@@ -82,69 +77,76 @@ Tensor empty_strided_cuda(IntArrayRef size, IntArrayRef stride, c10::optional<Sc
 
-  if (n < 30000) {  // For small inputs, we offload it to CPU instead.
-    auto result_cpu = at::empty({n}, result.options().device(kCPU));
-    randperm_out(result_cpu, n, generator);
-    return result.copy_(result_cpu);
+  auto range = at::arange(n, result.options());
+
+  // shuffled_data points to the underlying data of the output tensor if the tensor is contiguous; otherwise it
+  // points to a new tensor.
+  Tensor shuffled;
+  void *shuffled_data;
+  if (result.is_contiguous()) {
+    shuffled_data = result.data_ptr();
+  } else {
+    shuffled = at::empty(n, result.options());
+    shuffled_data = shuffled.data_ptr();
   }
 
-#if 0
-  // This if condition should never be true because if n >= 30000 and the tensor has a Half type,
-  // check_supported_max_int_with_precision should have reported an error. This snippet is commented out but left here
-  // for the sake of clarity, because Half in thrust is spotty, and we do not want future change unaware of this.
-  if (result.scalar_type() == at::ScalarType::Half) {  // Half in thrust is spotty. Avoid.
-    auto result_float = at::empty({n}, initialTensorOptions().device(Device(DeviceType::CUDA)));
-    return result.copy_(randperm_out_cuda(result_float, n, generator));
+  auto opt = TensorOptions().device(result.device());
+
+  int bits = std::min(64,
+    static_cast<int>(std::ceil(std::log2(static_cast<double>(n)))) + 1);
+
+  if (n == 0) {
+    return result;
+  } else if (bits <= 8) {
+    auto keys = at::empty(result.sizes(), opt.dtype(kByte)).random_(generator);
+    auto keys_tmp = at::empty_like(keys);
+    AT_DISPATCH_ALL_TYPES_AND(kHalf, result.scalar_type(), "randperm_out_cuda", [&] {
+      at::cuda::cub::sort_pairs<uint8_t, scalar_t>(
+        keys.data_ptr<uint8_t>(), keys_tmp.data_ptr<uint8_t>(),
+        range.data_ptr<scalar_t>(), reinterpret_cast<scalar_t*>(shuffled_data),
+        n, 0, bits);
+    });
+  } else if (bits <= 16) {
+    auto keys = at::empty(result.sizes(), opt.dtype(kShort)).random_(
+      std::numeric_limits<int16_t>::min(), std::numeric_limits<int16_t>::max(), generator);
+    auto keys_tmp = at::empty_like(keys);
+    AT_DISPATCH_ALL_TYPES_AND(kHalf, result.scalar_type(), "randperm_out_cuda", [&] {
+      at::cuda::cub::sort_pairs<int16_t, scalar_t>(
+        keys.data_ptr<int16_t>(), keys_tmp.data_ptr<int16_t>(),
+        range.data_ptr<scalar_t>(), reinterpret_cast<scalar_t*>(shuffled_data),
+        n, 0, bits);
+    });
+  } else if (bits <= 32) {
+    auto keys = at::empty(result.sizes(), opt.dtype(kInt)).random_(
+      std::numeric_limits<int>::min(), std::numeric_limits<int>::max(), generator);
+    auto keys_tmp = at::empty_like(keys);
+    AT_DISPATCH_ALL_TYPES(result.scalar_type(), "randperm_out_cuda", [&] {
+      at::cuda::cub::sort_pairs<int, scalar_t>(
+        keys.data_ptr<int>(), keys_tmp.data_ptr<int>(),
+        range.data_ptr<scalar_t>(), reinterpret_cast<scalar_t*>(shuffled_data),
+        n, 0, bits);
+    });
+  } else {
```

Comment:
this should never happen, given that you've errored out on n>int_max already?

Reply:
Ahh, yes!